test_binary_generation/generate_raw_ltoir.py

```diff
@@ -0,0 +1,178 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+
+import argparse
+import pathlib
+import subprocess
+import sys
+
+from cuda import nvrtc
+
+# Magic number found at the start of an LTO-IR file
+LTOIR_MAGIC = 0x7F4E43ED
+
+
+def check(args):
+    """
+    Abort and print an error message in the presence of an error result.
+
+    Otherwise:
+    - Return None if there were no more arguments,
+    - Return the singular argument if there was only one further argument,
+    - Return the tuple of arguments if multiple followed.
+    """
+
+    result, *args = args
+    value = result.value
+
+    if value:
+        error_string = check(nvrtc.nvrtcGetErrorString(result)).decode()
+        msg = f"NVRTC error, code {value}: {error_string}"
+        print(msg, file=sys.stderr)
+        sys.exit(1)
+
+    if len(args) == 0:
+        return None
+    elif len(args) == 1:
+        return args[0]
+    else:
+        return args
+
+
+def determine_include_path():
+    # Inspired by the logic in FindCUDA.cmake. We need the CUDA include path
+    # because NVRTC doesn't add it by default, and we can compile a much
+    # broader set of test files if the CUDA headers are available.
+
+    # We get NVCC to tell us the location of the CUDA Toolkit.
+
+    cmd = ["nvcc", "-v", "__dummy"]
```

Comment:
I'm not really sure - if you don't put something there, then `nvcc` refuses to print out its configuration, complaining of no input files.

Reply:
(I just followed what `FindCUDA.cmake` does)

Reply:
On my machine, even it prints out the config, it also prints this line:
```
nvcc fatal   : Don't know what to do with '__dummy'
```
So I'm not sure if it's actually effective. And I'm surprised it's not returning an error code.

Reply:
It does return an error code, which is 1 - I check to make sure it does that in the following lines because it's expected. If you omit the dummy argument then `nvcc` won't print its configuration.

Reply:
If this is also what FindCUDA does then it's probably ok. For reference, in Numbast, I made an empty tmp cu file and compile it with `-v` flag.

Reply:
I also tried compiling an empty file, with `nvcc -x cu -c -v -`, but that writes a bunch of stuff to disk, which I think is a bit inelegant.

I also tried preprocessing only with `nvcc -x cu -E -v -`, but that generates so much output I think I'd have to properly handle reading from stdout instead of just using `subprocess.run()` to avoid a potential deadlock (as per the warning in https://docs.python.org/3/library/subprocess.html#subprocess.Popen.stderr - I've hit this before working on other tooling).

Reply:
These kind of calls are used by CMake to scrape the nvcc output ( specifically the `#$` ) to get the CUDA toolkit binary directory, the implicit includes and libraries. We ( CMake ) also use `nvcc -c empty.cu -v` ( where empty.cu ) to get the full compile / link with the host compile details.